relay-server/src/utils/memory.rs

```diff
@@ -0,0 +1,269 @@
+            memory: ArcSwap::from(Arc::new(Self::refresh_memory(&mut system))),
+            last_update: AtomicU64::new(0),
+            reference_time: Instant::now(),
+            system: Mutex::new(system),
+        }))
+    }
+
+    /// Returns a copy of the most up to date [`Memory`] data.
+    pub fn memory(&self) -> Memory {
+        self.try_update();
+        **self.0.memory.load()
+    }
+
+    /// Builds a [`MemoryStatConfig`] which holds a reference to the supplied [`Config`] and the
+    /// current [`MemoryStat`].
+    pub fn with_config(&self, config: Arc<Config>) -> MemoryStatConfig {
+        MemoryStatConfig {
+            memory_stat: self.clone(),
+            config,
+        }
+    }
+
+    /// Refreshes the memory readings.
+    fn refresh_memory(system: &mut System) -> Memory {
+        system.refresh_memory_specifics(MemoryRefreshKind::new().with_ram());
+        let memory = match system.cgroup_limits() {
+            Some(cgroup) => Memory {
+                used: cgroup.rss,
+                total: cgroup.total_memory,
+            },
+            None => Memory {
+                used: system.used_memory(),
+                total: system.total_memory(),
+            },
+        };
+
+        metric!(gauge(RelayGauges::SystemMemoryUsed) = memory.used);
+        metric!(gauge(RelayGauges::SystemMemoryTotal) = memory.total);
+
+        memory
+    }
+
+    /// Updates the memory readings if at least [`UPDATE_TIME_THRESHOLD_MS`] has passed.
+    fn try_update(&self) {
+        let last_update = self.0.last_update.load(Ordering::Relaxed);
+        let elapsed_time = self.0.reference_time.elapsed().as_millis() as u64;
+
+        if elapsed_time - last_update < UPDATE_TIME_THRESHOLD_MS {
+            return;
+        }
+
+        let Ok(_) = self.0.last_update.compare_exchange_weak(
+            last_update,
+            elapsed_time,
+            Ordering::Relaxed,
+            Ordering::Relaxed,
+        ) else {
+            return;
+        };
+
+        let mut system = self
+            .0
+            .system
+            .lock()
+            .unwrap_or_else(|system| system.into_inner());
+
+        let updated_memory = Self::refresh_memory(&mut system);
+        self.0.memory.store(Arc::new(updated_memory));
+    }
+}
+
+impl fmt::Debug for MemoryStat {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "MemoryStat")
+    }
+}
+
+impl Default for MemoryStat {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// Struct that composes a [`Config`] and [`MemoryStat`] and provides utility methods to validate
+/// whether memory is within limits.
+///
+/// The rationale behind such struct, is to be able to share across Relay the same logic for dealing
+/// with memory readings. It's decoupled from [`MemoryStat`] because it's just a layer on top that
+/// decides how memory readings are interpreted.
+#[derive(Clone, Debug)]
+pub struct MemoryStatConfig {
```

Comment:
nit: Can we call this `MemoryChecker`? See also the `MemoryCheck` enum suggestion above.